File: geode-core/src/main/java/com/gemstone/gemfire/internal/cache/delta/Delta.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.gemstone.gemfire.internal.cache.delta;

import com.gemstone.gemfire.cache.EntryEvent;

/**
 * Represents changes to apply to a region entry instead of a new value.
 * A Delta is passed as the new value in a put operation on a Region
 * and knows how to apply itself to an old value.
 *
 * Internal Note: When an update message carries a Delta as a payload,
 * it makes sure it gets deserialized before being put into the region.
 *
 * @since GemFire 5.5
 * @see com.gemstone.gemfire.internal.cache.UpdateOperation
 */
public interface Delta {

  /**
   * Apply delta to the old value from the provided EntryEvent.
   * If the delta cannot be applied for any reason then an (unchecked)
   * exception should be thrown. If the put is being applied in a
   * distributed-ack scope, then the exception will be propagated back
   * to the originating put call, but will not necessarily cause puts
   * in other servers to fail.
   *
   * @param putEvent the EntryEvent for the put operation, from which
   * the old value can be obtained (as well as other information such
   * as the key and region being operated on)
   *
   * @return the new value to be put into the region
   */
  Object apply(EntryEvent<?, ?> putEvent);

  Object merge(Object toMerge, boolean isCreate);

  Object merge(Object toMerge);

  Object getResultantValue();
}

File: src/quickstart/csharp/Delta.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * The Delta QuickStart Example.
 *
 * This example takes the following steps:
 *
 * 1. Create a Geode Cache.
 * 2. Get the example Region from the Cache.
 * 3. Put an Entry into the Region.
 * 4. Set delta for a value.
 * 5. Put entry with delta in region.
 * 6. Local-invalidate entry in region.
 * 7. Get entry from server.
 * 8. Verify that delta was applied on server, by examining entry.
 * 9. Close the Cache.
 *
 */

// Use standard namespaces
using System;

// Use the Geode namespace
using Apache.Geode.Client;

namespace Apache.Geode.Client.QuickStart
{
  // The Delta QuickStart example.
  class Delta
  {
    static void Main(string[] args)
    {
      try
      {
        // Create a Geode Cache through XMLs/clientDelta.xml
        Properties<string, string> prop = Properties<string, string>.Create<string, string>();
        prop.Insert("cache-xml-file", "XMLs/clientDelta.xml");
        CacheFactory cacheFactory = CacheFactory.CreateCacheFactory(prop);
        Cache cache = cacheFactory.Create();

        Console.WriteLine("Created the Geode Cache");

        // Get the example Region from the Cache which is declared in the Cache XML file.
        IRegion<string, DeltaExample> region = cache.GetRegion<string, DeltaExample>("exampleRegion");

        Console.WriteLine("Obtained the Region from the Cache");

        Serializable.RegisterTypeGeneric(DeltaExample.create);

        //Creating Delta Object.
        DeltaExample ptr = new DeltaExample(10, 15, 20);
        
        //Put the delta object. This will send the complete object to the server.
        region["Key1"] = ptr;
        
        Console.WriteLine("Completed put for a delta object");


        //Changing state of delta object.
        ptr.setField1(9);

        //Put delta object again. Since delta flag is set true it will calculate
        //Delta and send only Delta to the server.
        region["Key1"] = ptr;
        Console.WriteLine("Completed put with delta");

        //Locally invalidating the key.
        region.GetLocalView().Invalidate("Key1");
        //Fetching the value from server.
        DeltaExample retVal = (DeltaExample) region["Key1"];

        //Verification
        if( retVal.getField1() != 9 )
          throw new GeodeException("First field should have been 9");
        if( retVal.getField2() != 15 )
          throw new GeodeException("Second field should have been 15");
        if( retVal.getField3() != 20 )
          throw new GeodeException("Third field should have been 20");
        
        Console.WriteLine("Delta has been successfully applied at server");

        // Close the Geode Cache.
        cache.Close();

        Console.WriteLine("Closed the Geode Cache");

      }
      // An exception should not occur
      catch (GeodeException gfex)
      {
        Console.WriteLine("Delta Geode Exception: {0}", gfex.Message);
      }
    }
  }
}
